File: miscellaneous/Miscellaneous/Tutorials/2009-EDSys/Examples/C#/SchedSquare.cs
// -------------------------------------------------------------- -*- C# -*-
// File: examples/src/csharp/SchedSquare.cs
// --------------------------------------------------------------------------
// Copyright (C) 1990-2008 by ILOG.
// All Rights Reserved.
//
// Permission is expressly granted to use this example in the
// course of developing applications that use ILOG products.
// --------------------------------------------------------------------------

/* ------------------------------------------------------------

Problem Description
-------------------

The aim of the square example is to place a set of small squares of
different sizes into a large square.

------------------------------------------------------------ */

using System;
using ILOG.CP;
using ILOG.Concert;

namespace SchedSquare
{
    public class SchedSquare
    {
        public static void Main(String[] args)
        {
            try
            {
                CP cp = new CP();

                int sizeSquare = 112;
                int nbSquares = 21;
                int[] size = { 50, 42, 37, 35, 33, 29, 27, 25, 24, 19, 18, 17, 16, 15, 11, 9, 8, 7, 6, 4, 2 };
                IIntervalVar[] x = new IIntervalVar[nbSquares];
                IIntervalVar[] y = new IIntervalVar[nbSquares];
                ICumulFunctionExpr rx = cp.CumulFunctionExpr();
                ICumulFunctionExpr ry = cp.CumulFunctionExpr();

                for (int i = 0; i < nbSquares; ++i)
                {
                    x[i] = cp.IntervalVar(size[i], "X" + i);
                    x[i].EndMax = sizeSquare;
                    y[i] = cp.IntervalVar(size[i], "Y" + i);
                    y[i].EndMax = sizeSquare;
                    rx = cp.Sum(rx, cp.Pulse(x[i], size[i]));
                    ry = cp.Sum(ry, cp.Pulse(y[i], size[i]));

                    for (int j = 0; j < i; ++j)
                    {
                        cp.Add(cp.Or(cp.Le(cp.EndOf(x[i]), cp.StartOf(x[j])),
                                cp.Or(cp.Le(cp.EndOf(x[j]), cp.StartOf(x[i])),
                                        cp.Or(cp.Le(cp.EndOf(y[i]), cp.StartOf(y[j])),
                                                cp.Le(cp.EndOf(y[j]), cp.StartOf(y[i]))))));
                    }
                }
                cp.Add(cp.AlwaysIn(rx, 0, sizeSquare, sizeSquare, sizeSquare));
                cp.Add(cp.AlwaysIn(ry, 0, sizeSquare, sizeSquare, sizeSquare));

                ISearchPhase[] phases = new ISearchPhase[2];
                phases[0] = cp.SearchPhase(x);
                phases[1] = cp.SearchPhase(y);

                if (cp.Solve(phases))
                {
                    for (int i = 0; i < nbSquares; ++i)
                    {
                        Console.WriteLine("Square " + i + ": ["
                                + cp.GetStart(x[i]) + "," + cp.GetEnd(x[i])
                                + "] x ["
                                + cp.GetStart(y[i]) + "," + cp.GetEnd(y[i])
                                + "]");
                    }
                }
            }
            catch (IloException ex)
            {
                Console.WriteLine("Caught: " + ex);
            }
        }
    }
}

File: miscellaneous/Miscellaneous/Tutorials/2009-EDSys/Examples/Java/SchedSquare.java
// -------------------------------------------------------------- -*- Java -*-
// File: examples/src/java/SchedSquare.java
// --------------------------------------------------------------------------
// Copyright (C) 1990-2008 by ILOG.
// All Rights Reserved.
//
// Permission is expressly granted to use this example in the
// course of developing applications that use ILOG products.
// --------------------------------------------------------------------------

/* ------------------------------------------------------------

Problem Description
-------------------

The aim of the square example is to place a set of small squares of
different sizes into a large square.

------------------------------------------------------------ */

//$doc:ALL
//$doc:HEADERS

import ilog.concert.*;
import ilog.cp.*;

import java.util.List;
import java.util.ArrayList;

//end:HEADERS

public class SchedSquare {

    //$doc:MAIN
    public static void main(String[] args) {

        try {
//end:MAIN
            IloCP cp = new IloCP();

            int sizeSquare = 112;
            int nbSquares = 21;
            int[] size = {50, 42, 37, 35, 33, 29, 27, 25, 24, 19, 18, 17, 16, 15, 11, 9, 8, 7, 6, 4, 2};
            IloIntervalVar[] x = new IloIntervalVar[nbSquares];
            IloIntervalVar[] y = new IloIntervalVar[nbSquares];
            IloCumulFunctionExpr rx = cp.cumulFunctionExpr();
            IloCumulFunctionExpr ry = cp.cumulFunctionExpr();

            for (int i = 0; i < nbSquares; ++i) {
                x[i] = cp.intervalVar(size[i], "X" + i);
                x[i].setEndMax(sizeSquare);
                y[i] = cp.intervalVar(size[i], "Y" + i);
                y[i].setEndMax(sizeSquare);
                rx = cp.sum(rx, cp.pulse(x[i], size[i]));
                ry = cp.sum(ry, cp.pulse(y[i], size[i]));

                for (int j = 0; j < i; ++j) {
                    cp.add(cp.or(cp.le(cp.endOf(x[i]), cp.startOf(x[j])),
                            cp.or(cp.le(cp.endOf(x[j]), cp.startOf(x[i])),
                                    cp.or(cp.le(cp.endOf(y[i]), cp.startOf(y[j])),
                                            cp.le(cp.endOf(y[j]), cp.startOf(y[i]))))));
                }
            }
            cp.add(cp.alwaysIn(rx, 0, sizeSquare, sizeSquare, sizeSquare));
            cp.add(cp.alwaysIn(ry, 0, sizeSquare, sizeSquare, sizeSquare));

            IloSearchPhase[] phases = new IloSearchPhase[2];
            phases[0] = cp.searchPhase(x);
            phases[1] = cp.searchPhase(y);

            if (cp.solve(phases)) {
                for (int i = 0; i < nbSquares; ++i) {
                    System.out.println("Square " + i + ": ["
                            + cp.getStart(x[i]) + "," + cp.getEnd(x[i])
                            + "] x ["
                            + cp.getStart(y[i]) + "," + cp.getEnd(y[i])
                            + "]");
                }
            }

        } catch (IloException ex) {
            System.out.println("Caught: " + ex);
        }
    }
}
//$end:ALL
